### ISSUE TITLE
OSError: Too many files open  

### ISSUE THREAD

**Issue by cdyrssen** [madmaze/pytesseract#209]
I have been using this python wrapper for tesseract on a video feed with opencv in order to extract time series data into csv files. I am trying to track 40 different windows. However, on the 201st frame of the test video I have, I get the following exception:
```
Traceback (most recent call last):
  File "/usr/lib/python3.7/site-packages/pytesseract/pytesseract.py", line 219, in run_tesseract
    proc = subprocess.Popen(cmd_args, **subprocess_args())
  File "/usr/lib/python3.7/subprocess.py", line 775, in __init__
    restore_signals, start_new_session)
  File "/usr/lib/python3.7/subprocess.py", line 1412, in _execute_child
    errpipe_read, errpipe_write = os.pipe()
OSError: [Errno 24] Too many open files

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "bars_and_windows.py", line 54, in <module>
    bar_text = pytesseract.image_to_string(tmp_img, lang='SF_Upper')
  File "/usr/lib/python3.7/site-packages/pytesseract/pytesseract.py", line 341, in image_to_string
    }[output_type]()
  File "/usr/lib/python3.7/site-packages/pytesseract/pytesseract.py", line 340, in <lambda>
    Output.STRING: lambda: run_and_get_output(*args),
  File "/usr/lib/python3.7/site-packages/pytesseract/pytesseract.py", line 249, in run_and_get_output
    run_tesseract(**kwargs)
  File "/usr/lib/python3.7/site-packages/pytesseract/pytesseract.py", line 221, in run_tesseract
    raise TesseractNotFoundError()
pytesseract.pytesseract.TesseractNotFoundError: tesseract is not installed or it's not in your path
```

I was originally writing to 40 different csv files simultaneously. After completely removing my file writing/reading code, this error still persisted and triggered on the exact same frame of the video every time. I also tested with a different video and ran into the same exception after only 25 frames. This leads me to believe this is a bug/limitation in pytesseract. Has anyone else run into a similar issue and been able to come up with a solution/work around?

My setup:
Linux Kernel: 5.1.16-arch1-1-ARCH
python-pytesseract-git package installed from aur
Python 3.7.3-2

**Comment by int3l** [comment]
Hi @cdyrssen 
thanks for reporting the issue.
I can see that the [python-pytesseract-git](https://aur.archlinux.org/packages/python-pytesseract-git/) package is a bit old (from 2018) and the version there is 0.2.6. The version itself is not the main problem, but it is a good idea to update to the latest version (0.2.7 or soon to 0.2.8 from git).

About the issue itself: your python process (that runs **bars_and_windows.py**) reaches the limit of files that it can open at the same time from your current user. This is a normal policy in the Linux world.
Maybe your script does parallel processing over multiple frames (images) at once, right?

Can you provide example snippet of code that triggers the exception?
Also please check what is the limit for your current user with the command `ulimit -n` in your terminal.

**Comment by cdyrssen** [comment]
@int3l Thanks for the response!  Here is the python script I am using at the moment. The interesting lines are 52-66.

```
from time import time
from PIL import Image
import numpy as np
import pytesseract
import resource
import cv2
import os

limit = resource.getrlimit(resource.RLIMIT_NOFILE)
resource.setrlimit(resource.RLIMIT_NOFILE, limit)

cap = cv2.VideoCapture('July10.mp4')
cv2.namedWindow('frame', cv2.WND_PROP_FULLSCREEN)
cv2.setWindowProperty('frame', cv2.WND_PROP_FULLSCREEN, cv2.WINDOW_FULLSCREEN)

# Grid layout
windows = (8, 5)

# Global traversal values
x_inc, y_inc = 512, 427
adj_mtx = np.zeros(windows+(2,), dtype=int)
adj_mtx[:, -1, 0] = 2
adj_mtx[:, 2, 1] = 13
adj_mtx[:, 3:, 1] = 45

# Bar dimensions
x_bar_orig, y_bar_orig = 1500, 38
x_bar_off, y_bar_off = 1075, 31

# Window dimensions
x_win_orig, y_win_orig = 4, 153
x_win_off, y_win_off = 461, 270

scale_factor = 1
height = int(y_bar_off*scale_factor)
width = int(x_bar_off*scale_factor)
dim = (width, height)

start = time()
frame_cnt = 1
while(cap.isOpened()):
    ret, frame = cap.read()

    i = 0
    err_cnt = 0
    x_bar_pt1, y_bar_pt1 = x_bar_orig, y_bar_orig
    x_bar_pt2 = x_bar_orig + x_bar_off
    for j in range(windows[1]):
        y_bar_pt1 = y_bar_orig + j*y_inc + adj_mtx[i, j, 1]
        y_bar_pt2 = y_bar_pt1+y_bar_off

        # Process bar sub-image
        bar = cv2.cvtColor(frame[y_bar_pt1:y_bar_pt2, x_bar_pt1:x_bar_pt2],
                           cv2.COLOR_BGR2GRAY)
        bar = cv2.resize(bar, dim, interpolation=cv2.INTER_LINEAR)
        tmp_img = Image.fromarray(bar)
        tmp_img.format = 'PNG'

        # Extract text from bar sub-image
        bar_text = pytesseract.image_to_string(tmp_img, lang='SF_Upper')
        bar_text = bar_text.split('-')[0]
        if bar_text == '':
            cv2.imshow('fail{}'.format(err_cnt), bar)
            err_cnt += 1
        else:
            cv2.imwrite('{}.png'.format(bar_text), bar)

        symbols = bar_text.split(',')
        y_win_pt1 = y_win_orig + j*y_inc + adj_mtx[i, j, 1]
        if len(symbols) == windows[0]:
            for i in range(windows[0]):
                x_win_pt1= x_win_orig + i*x_inc + adj_mtx[i, j, 0]
                x_win_pt2, y_win_pt2 = x_win_pt1+x_win_off, y_win_pt1+y_win_off

                win_coord1 = (x_win_pt1, y_win_pt1)
                win_coord2 = (x_win_pt2, y_win_pt2)
                cv2.rectangle(frame, win_coord1, win_coord2, (0, 0, 255), 1)

                symbol = symbols[i].replace(' ', '')
                symbol_file = '{}.csv'.format(symbol)
                with open(symbol_file, 'w') as csv:
                    csv.write(bar_text)
                    csv.close()

        # Display bar boxes
        bar_coord1 = (x_bar_pt1, y_bar_pt1)
        bar_coord2 = (x_bar_pt2, y_bar_pt2)
        cv2.rectangle(frame, bar_coord1, bar_coord2, (255, 0, 0), 1)
    cv2.imshow('frame', frame)
    print('new frame {}'.format(frame_cnt))
    frame_cnt += 1

    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

cap.release()
cv2.destroyAllWindows()
```

Basically, it loads a screen capture video file and goes through each frame serially. Within each frame I have 40 areas I am trying to track (5 program windows with 8 sub-windows). Each area will have its own csv file associated with it. The goal is to expand on this code to watch for new data to come into view in each area and extract that data via tesseract-ocr into the appropriate csv file.

I've done a little research and came across the ulimit command and upped my limit from 1024 to 8192 (arch wiki suggests this number for some video games to work properly). However, this just caused the issue to occur at a later frame. I tried modifying the pytesseract code to include the 'close_fds=True' argument within the subprocess.Popen() call of the run_tesseract() function. This didn't seem to fix the issue either. It causes the exception at the same frame. Evidently, this is a known issue when working with many calls to subprocess.Popen(). I'm wondering if there is alternative that can be used instead?

In the mean time I will be updating to the latest version of the code and looking a further into this issue. It seems like files or pipes are not being closed when the subprocess completes.

**Comment by int3l** [comment]
Hmm, very interesting. I can't reproduce this locally.
I have the following snippet of code:

```python
import pytesseract
from PIL import Image
import cv2


image_path = 'example.jpg'
tmp = cv2.imread(image_path)
while True:
        image = Image.fromarray(tmp)
        image.format = 'PNG'
        pytesseract.image_to_string(image)
```

I monitor the open files with `watch "lsof -p <pid> | wc -l"` and there is no new descriptors.
So there must be something specific between opencv and pytesseract.
Do you get the same error If you replace the line with **image_to_string** to something like `bar_text = "test-"`?

**Comment by cdyrssen** [comment]
Modifying the following lines:
```
        # Extract stock symbols from bar sub-image
        bar_text = pytesseract.image_to_string(tmp_img, lang='SF_Upper')
        bar_text = bar_text.split('-')[0]
        if bar_text == '':
            cv2.imshow('fail{}'.format(err_cnt), bar)
            err_cnt += 1
        else:
            cv2.imwrite('{}.png'.format(bar_text), bar)
```
to:
```
        # Extract stock symbols from bar sub-image
       bar_text = ''.join(random.choice(string.ascii_lowercase) for i in range(6))
        bar_text = bar_text.split('-')[0]
        if bar_text == '':
            cv2.imshow('fail{}'.format(err_cnt), bar)
            err_cnt += 1
        else:
            cv2.imwrite('{}.png'.format(bar_text), bar)
```
did not trigger the exception.

**Comment by int3l** [comment]
So it is indeed the pytesseract itself. Ok, I will try the pytesseract package from the Aur repo.

**Comment by cdyrssen** [comment]
Also after running "pip3 search pytesseract" it says my installed version is 0.2.8 and the latest version available to pip is 0.2.7.

**Comment by int3l** [comment]
This is because you did the install from git master :)
And 0.2.8 is not released to PyPI.

**Comment by cdyrssen** [comment]
Found [this](https://gist.github.com/murrant/4db51c851db480abb6c36a308d3bcef1) snippet of code and put modified it a little and put in inside of a try-except block like this:

```
        # Extract stock symbols from bar sub-image
        bar_text = ''
        try:
            bar_text = pytesseract.image_to_string(tmp_img, lang='SF_Upper')
        except:
            import os
            import inspect
            descriptors = set()
            (frame, filename, line_number, function_name, lines, index) = inspect.getouterframes(inspect.currentframe())[0]
            fds = set(os.listdir('/proc/self/fd/'))
            new_fds = fds - descriptors
            closed_fds = descriptors - fds
            descriptors = fds
            print("{}:{} ALL file descriptors: {}".format(filename, line_number, fds))
            print("{}:{} NEW file descriptors: {}".format(filename, line_number, new_fds))
            print("{}:{} CLOSED file descriptors: {}".format(filename, line_number, closed_fds))
            exit()
        # bar_text = ''.join(random.choice(string.ascii_lowercase) for i in range(6))
        bar_text = bar_text.split('-')[0]
        if bar_text == '':
            cv2.imshow('fail{}'.format(err_cnt), bar)
            err_cnt += 1
        else:
            cv2.imwrite('{}.png'.format(bar_text), bar)
```

and get output like this:
```
bars_and_windows.py:66 ALL file descriptors: {'247', '213', '452', '437', '329', '1', '505', '51', '435', '325', '60', '315', '443', '422', '77', '64', '301', '463', '474', '19', '119', '270', '274', '385', '180', '202', '455', '210', '473', '127', '131', '406', '432', '122', '69', '58', '205', '323', '490', '402', '170', '160', '228', '487', '346', '166', '144', '431', '499', '405', '414', '282', '312', '36', '314', '104', '182', '34', '229', '133', '148', '373', '102', '187', '137', '245', '349', '380', '404', '400', '39', '461', '458', '465', '485', '222', '49', '420', '72', '287', '502', '268', '419', '337', '99', '497', '28', '353', '257', '339', '342', '415', '207', '240', '136', '243', '281', '338', '426', '42', '204', '477', '162', '139', '311', '269', '115', '178', '283', '288', '165', '456', '433', '159', '252', '333', '409', '271', '258', '152', '302', '482', '10', '188', '190', '378', '498', '489', '20', '0', '146', '211', '366', '324', '438', '189', '3', '179', '290', '198', '403', '176', '35', '316', '399', '292', '107', '254', '374', '244', '304', '111', '381', '7', '256', '383', '251', '390', '371', '232', '320', '241', '145', '63', '90', '5', '478', '344', '33', '65', '70', '6', '354', '501', '26', '445', '223', '80', '235', '249', '47', '397', '157', '172', '67', '279', '181', '24', '436', '110', '469', '297', '30', '412', '343', '242', '429', '195', '470', '430', '32', '224', '298', '460', '365', '272', '185', '112', '94', '106', '149', '389', '151', '13', '464', '295', '503', '277', '495', '237', '358', '217', '66', '57', '155', '363', '444', '8', '310', '143', '321', '453', '50', '203', '416', '442', '225', '73', '74', '193', '16', '305', '263', '500', '116', '220', '336', '355', '31', '169', '449', '423', '233', '491', '382', '476', '130', '14', '141', '296', '275', '307', '209', '86', '418', '362', '12', '132', '410', '2', '117', '234', '267', '266', '291', '386', '41', '475', '199', '459', '62', '215', '48', '114', '326', '59', '121', '23', '434', '147', '303', '118', '153', '348', '89', '322', '439', '368', '425', '424', '375', '472', '273', '335', '327', '411', '484', '471', '21', '9', '79', '196', '150', '154', '191', '22', '25', '177', '300', '259', '294', '309', '407', '408', '255', '393', '391', '98', '197', '97', '219', '488', '261', '105', '95', '214', '125', '351', '123', '140', '356', '91', '174', '285', '496', '454', '441', '299', '450', '483', '357', '226', '446', '448', '396', '457', '218', '308', '372', '192', '100', '367', '43', '171', '248', '120', '96', '250', '236', '238', '401', '227', '421', '350', '85', '486', '183', '156', '208', '52', '68', '126', '377', '46', '168', '428', '128', '284', '29', '37', '440', '103', '109', '276', '101', '194', '38', '313', '330', '332', '18', '392', '124', '83', '44', '253', '379', '138', '394', '4', '278', '319', '76', '158', '184', '265', '376', '479', '17', '93', '113', '142', '239', '161', '384', '262', '230', '246', '494', '318', '92', '212', '54', '173', '82', '447', '175', '221', '480', '200', '388', '369', '45', '359', '360', '504', '56', '164', '81', '75', '398', '27', '280', '15', '467', '40', '260', '286', '466', '61', '347', '53', '186', '364', '387', '361', '163', '108', '331', '395', '129', '462', '201', '206', '413', '493', '340', '71', '468', '84', '216', '341', '88', '87', '167', '352', '55', '328', '481', '370', '135', '317', '306', '492', '264', '134', '78', '231', '334', '417', '345', '293', '427', '289', '11', '451'}
bars_and_windows.py:66 NEW file descriptors: {'247', '213', '296', '452', '437', '275', '329', '1', '505', '307', '209', '51', '86', '435', '418', '325', '60', '362', '315', '443', '12', '132', '422', '410', '2', '77', '117', '234', '64', '267', '301', '266', '463', '291', '386', '474', '19', '119', '270', '274', '41', '385', '475', '180', '202', '199', '459', '455', '210', '62', '215', '473', '48', '114', '326', '59', '121', '23', '434', '127', '147', '131', '406', '303', '432', '118', '153', '348', '89', '122', '69', '322', '58', '439', '205', '368', '425', '323', '490', '402', '170', '424', '160', '375', '472', '228', '487', '273', '346', '335', '327', '166', '144', '431', '411', '484', '499', '471', '21', '405', '414', '282', '9', '79', '312', '36', '314', '104', '182', '34', '229', '133', '148', '196', '150', '373', '154', '102', '191', '187', '22', '137', '25', '245', '177', '300', '259', '349', '380', '404', '400', '294', '309', '39', '407', '408', '461', '458', '255', '393', '391', '98', '197', '97', '465', '485', '219', '488', '222', '49', '420', '261', '105', '72', '95', '287', '214', '125', '351', '123', '140', '356', '91', '174', '285', '496', '454', '441', '299', '450', '483', '502', '357', '268', '419', '226', '446', '337', '448', '396', '99', '497', '28', '353', '257', '339', '457', '218', '308', '342', '415', '372', '192', '207', '240', '136', '243', '100', '367', '43', '171', '281', '338', '426', '42', '248', '120', '204', '96', '250', '236', '477', '162', '238', '139', '311', '401', '269', '227', '115', '421', '178', '283', '350', '288', '85', '486', '183', '156', '208', '165', '456', '52', '68', '126', '377', '433', '46', '168', '159', '428', '128', '252', '284', '333', '409', '29', '271', '37', '258', '440', '152', '103', '302', '482', '109', '276', '101', '10', '188', '190', '378', '498', '194', '489', '20', '0', '146', '211', '366', '38', '313', '324', '330', '332', '18', '392', '438', '189', '124', '83', '44', '253', '379', '138', '3', '179', '290', '394', '198', '403', '4', '278', '176', '35', '316', '399', '319', '292', '76', '158', '184', '107', '265', '254', '374', '376', '244', '304', '479', '111', '17', '381', '93', '7', '113', '256', '383', '251', '390', '371', '142', '239', '161', '384', '262', '230', '232', '320', '241', '246', '494', '318', '145', '92', '212', '54', '173', '63', '90', '82', '5', '478', '447', '344', '33', '65', '70', '175', '6', '354', '501', '26', '221', '445', '480', '200', '223', '80', '388', '369', '235', '45', '359', '360', '249', '47', '397', '504', '56', '157', '172', '67', '164', '279', '181', '81', '24', '436', '110', '469', '297', '30', '412', '343', '75', '398', '242', '27', '429', '195', '470', '280', '430', '15', '467', '32', '40', '260', '286', '466', '61', '224', '298', '460', '365', '347', '53', '186', '272', '364', '387', '185', '361', '163', '112', '108', '331', '94', '106', '395', '462', '149', '389', '151', '13', '201', '464', '206', '413', '295', '503', '493', '277', '340', '495', '237', '71', '358', '217', '468', '66', '57', '155', '363', '84', '216', '341', '444', '8', '310', '143', '321', '453', '88', '50', '87', '203', '416', '167', '352', '55', '442', '225', '328', '481', '73', '74', '193', '16', '370', '135', '305', '317', '263', '500', '116', '306', '492', '220', '336', '264', '134', '78', '231', '355', '289', '31', '169', '334', '449', '417', '423', '345', '233', '491', '382', '476', '130', '14', '293', '427', '129', '11', '451', '141'}
bars_and_windows.py:66 CLOSED file descriptors: set()
```

**Comment by cdyrssen** [comment]
I was also able to reproduce the bug with your snippet code:
```
import pytesseract
from PIL import Image
import cv2


image_path = 'example.jpg'
tmp = cv2.imread(image_path)
while True:
        image = Image.fromarray(tmp)
        image.format = 'PNG'
        pytesseract.image_to_string(image)
```
It occurred after 1012 iterations with my ulimit -Sn set to 512.

**Comment by int3l** [comment]
I will try to use the same python version as you, because this is the only difference.

**Comment by cdyrssen** [comment]
Opencv version: 4.1.0.25 installed via pip3
Pillow version: 6.1.0-1 installed via pacman

**Comment by int3l** [comment]
Sigh... I still cant reproduce with:
Ubuntu 18.04.2 LTS
Python 3.7.3-2~18.04.1
Tesseract 4.00~git2288-10f4998a-2 (tesseract 4.0.0-beta.1)
OpenCV 3.2.0+dfsg-4ubuntu0.1

Pip:
numpy 1.16.4
opencv-python 4.1.0.25
Pillow 6.1.0
pytesseract 0.2.6 or 0.2.8

What are the versions of your Tesseract and OpenCV lib/binary packages?

**Comment by cdyrssen** [comment]
tesseract 4.1.0-1
opencv 4.1.0-2

The weird thing is I don't remember running into this issue when I first started this project. I don't recall if I just didn't let it run for the entire duration before closing it out or if it actually went through the entire video without issues before..

**Comment by cdyrssen** [comment]
Downgraded to:
tesseract 4.0.0-2
opencv 3.4.2-2 (oldest version available in arch packages archive)

Still same result. I will try running this code on my desktop to see if I get the same error. Not sure why my laptop would cause issues though if it works on another machine..

Laptop: Dell Inspiron 5577 running Arch - Gnome/I3 (Xorg no Wayland support)

**Comment by int3l** [comment]
Let me build an arch image. Because I am running out of ideas with Ubuntu dev setup :D

**Comment by cdyrssen** [comment]
Interesting development.. After creating a virtualenv and installing all the packages (Pillow, opencv, and pytesseract) via pip3, this exception no longer seems to be occurring. I will continue development using this virtual env (probably a good idea anyways). This may not be pytesseract's fault after all and might be an issue with the health of my local python environment.

**Comment by int3l** [comment]
Yea, very strange 🔢 I managed to build the archlinux/base image with all the deps that you specified and when I run both tests (your and mine example), I still couldn't trigger exception and I don't observe increase of open file descriptors.

Feel free to close the issue once you are sure that it was a local dev environment issue.

**Comment by cdyrssen** [comment]
Yep I verified with `watch "lsof -p <pid> | wc -l"` command the number of open file descriptors stayed relatively consistent (anywhere from 93-97). Whenever I run the program outside of the virtual environment this number always increases until the exception occurs. Not sure what causes this in my local python environment, but I cannot reproduce on my desktop or in virtual environment. Will have to blame local dev environment. Thanks for your help! Closing this issue now.
